enterprise/server/remote_execution/commandutil/commandutil.go

```diff
@@ -136,25 +157,65 @@ func Run(ctx context.Context, command *repb.Command, workDir string, opts *conta
 //
 // For an example command that can be passed to this func, see
 // constructExecCommand.
-func RunWithProcessTreeCleanup(ctx context.Context, cmd *exec.Cmd) error {
+func RunWithProcessTreeCleanup(ctx context.Context, cmd *exec.Cmd, opts *RunWithCleanupOpts) (*container.Stats, error) {
 	if err := cmd.Start(); err != nil {
-		return err
+		return nil, err
 	}
 	pid := cmd.Process.Pid
-	done := make(chan struct{})
-	defer close(done)
+	processTerminated := make(chan struct{})
+	wait := func() error {
+		defer close(processTerminated)
+		return cmd.Wait()
+	}
+	// Cleanup goroutine: kill the process tree when the context is canceled.
 	go func() {
 		select {
-		case <-done:
+		case <-processTerminated:
 			return
 		case <-ctx.Done():
 			if err := KillProcessTree(pid); err != nil {
 				log.Warningf("Failed to kill process tree: %s", err)
 			}
 		}
 	}()
+	statsCh := make(chan *container.Stats, 1)
+	// Monitor goroutine: periodically record process stats.
+	go func() {
+		defer close(statsCh)
+		if !opts.EnableStats {
+			return
+		}
+		statsCh <- collectStats(pid, processTerminated)
+	}()
 
-	return cmd.Wait()
+	if err := wait(); err != nil {
+		return nil, err
+	}
+	return <-statsCh, nil
+}
+
+func collectStats(pid int, done <-chan struct{}) *container.Stats {
+	ts := NewTreeStats(pid)
+	// Most processes are short-lived so we need a fast poll rate if we want
+	// to increase the probability of getting at least one sample. But
+	// polling is expensive: a 50ms poll rate consumes around 10% of a CPU
+	// on Linux (and it's even worse on macOS). So we take a hybrid approach
+	// here: start off polling fast but slow down over time.
+	pollInterval := 25 * time.Millisecond
+	const pollBackoff = 1.25
+	const maxPollInterval = 1000 * time.Millisecond
```

Comment:
@tylerwilliams I couldn't beat the performance of `gopsutil` even using the raw `procfs` (looking at `/proc/[pid]/stat`), so I came up with this compromise for reducing CPU overhead. LMK what you think